### PR TITLE
refactor(source): prefer `SpecificParserConfig` over `SourceStruct`

### PR DESCRIPTION
**Pull request by xiangjinwu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Preparation for the following goal:
* consolidate `enum ProtocolProperties` and `enum SourceFormat` into a single `enum ConnectorFormat` (or derived similar to encode below when needed).
* consolidate `enum EncodingProperties` and `enum SourceEncode` into `enum ConnectorEncode` with derived [`EnumDiscriminants`](https://docs.rs/strum_macros/0.25.2/strum_macros/derive.EnumDiscriminants.html).

In this first part, we observe that most callers are always calling these two in sequence:
* `extract_source_struct: PbStreamSourceInfo -> SourceStruct`
* `SpecificParserConfig::new: (SourceStruct, PbStreamSourceInfo, HashMap<String, String>) -> Self`

So we change it into a single call, not exposing the intermediate `SourceStruct`:
* `SpecificParserConfig::new: (PbStreamSourceInfo, HashMap<String, String>) -> Self`

Other places using `SourceStruct`:
* `FsSourceDesc` does not use `source_struct`. It already contains `SpecificParserConfig` inside `FsConnectorSource`.
* `DatagenEventGenerator` only supports json today. When it supports more encodings later, it would need the properties and having only enum discriminants is not enough.
* `handle_alter_source_column` rejects avro or protobuf based on discriminant alone. It also rejects json with schema registry. It is unclear whether we need discriminant alone or the full properties. This usage has not been changed by this PR.

In the second part, we will do trivial renames to consolidate the enums as mentioned above.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/handler/create_source.rs [hunk 1]
```diff
@@ -99,11 +99,7 @@ async fn extract_avro_table_schema(
     info: &StreamSourceInfo,
     with_properties: &HashMap<String, String>,
 ) -> Result<Vec<ColumnCatalog>> {
-    let parser_config = SpecificParserConfig::new(
-        SourceStruct::new(SourceFormat::Plain, SourceEncode::Avro),
-        info,
-        with_properties,
-    )?;
+    let parser_config = SpecificParserConfig::new(info, with_properties)?;
```
Comment (xiangjinwu) [inline src/frontend/src/handler/create_source.rs:102]: One of the caller is actually passing in UpsertAvro today. Is the caller supposed to call `extract_upsert_avro_table_schema`?

It does not affect correctness right now because only `encoding_properties` field is used, and `SpecificParserConfig::new` handles UpsertAvro and PlainAvro the same for encoding.

https://github.com/risingwavelabs/risingwave/blob/6fb7ae5f47addc3953e3be76778c70dd611a5bf4/src/frontend/src/handler/create_source.rs#L532-L559

Comment (xiangjinwu) [reply src/frontend/src/handler/create_source.rs:102]: Dates back to first day upsert was supported: #8111

**Review by fuyufjh — APPROVED** [review]
The motivation LGTM

**Review by tabVersion — APPROVED** [review]
overall lgtm

**Comment by xiangjinwu** [comment]
Let's hold this PR for a bit until we figure out whether the inconsistency above is intentional.

**Comment by xiangjinwu** [comment]
According to a recent offline discussion, the special path of upsert-avro-with-sql_defined_pk pretending to be plain-avro is going to be removed. This refactor will be rebased after that is done.

**Comment by xiangjinwu** [comment]
PR merged after digging and confirming an answer for the following question:

> One of the caller is actually passing in UpsertAvro today. Is the caller supposed to call `extract_upsert_avro_table_schema`?

TLDR:
* Before this PR, the relevant code path allows certain `create source/table` to succeed, but the source parser constantly throws errors that are suppressed.
* After this PR, such `create source/table` would fail with the corresponding source parser error.

Back to the question, there were 3 options:
1. (Before this PR) call `extract_avro_table_schema`, which pretends to be `(plain, avro)`.
2. (After this PR) call `extract_avro_table_schema`, which uses `(upsert, avro)` from passed `source_info`.
3. (Question above) call `extract_upsert_avro_table_schema`.

The call was made by upsert avro with user defined primary key in sql.
* `extract_upsert_avro_table_schema` is essentially `extract_avro_table_schema` followed by key validation/mutation
  * When key schema is defined, it ensures key columns are subset of value columns.
  * When key schema is not defined, it adds an `_rw_key` column in addition to value columns.
  * This extra behavior is not intended to be followed by the upsert-avro-with-user-defined-parimay-key-in-sql case. So the 3rd option is not what we want.
* `SpecificParserConfig::new` does handle upsert avro and plain avro differently.
  * On `upsert`, it requires use of schema registry, and always fetches both key and value schema[1].
  * On `plain`, it allows both schema registry and file/https/s3 URLs[2]. It only fetches value schema.
  * Note that when using schema registry with key schema in schema registry, there is essentially no difference.
* `SpecificParserConfig::new` is called at least twice, one in the frontend and one in the compute source executor.
  * The frontend call is our focus here, which either (1) pretends to be `(plain, avro)` or (2) uses `(upsert, avro)` from passed `source_info`.
  * The compute source executor call has always been using `(upsert, avro)` from passed `source_info`.

As a result, in cases where `SpecificParserConfig::new` does behave differently, for example not using schema registry[2], or key schema is not in schema registry[1], the frontend call would succeed but the compute source executor call would fail. No data would be ingested and it will retry with warnings forever[3]. By changing from (1) to (2) in PR, we are able to report such error in frontend.

[1] Schema registry does allow key schema to be absent. In most parts of doc it uses `StringSerializer` (rather than avro/proto/json) for key. Requiring key schema, especially when user defines primary key in sql, might be a bug.
[2] Current implementation of avro without schema registry is not practical due to #12871
[3] We tolerate errors in source ([src](https://github.com/risingwavelabs/risingwave/blob/bbd28521a296c007c1b761f8011ffc50c134cdcc/src/stream/src/executor/source/source_executor.rs#L449-L460)). However, the actual error was not included in warning message, but used as a metrics label. This makes the warning less useful, and metrics cardinality potentially huge.
